test/e2e/kcp_remediations.go

```diff
@@ -0,0 +1,693 @@
+		thirdMachineReplacement := &clusterv1.Machine{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      thirdMachineReplacementName,
+				Namespace: namespace.Name,
+			},
+		}
+		Expect(input.BootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKeyFromObject(thirdMachineReplacement), thirdMachineReplacement)).To(Succeed(), "Failed to get machine %d", thirdMachineReplacementName)
+		Expect(thirdMachineReplacement.Status.NodeRef).To(BeNil())
+		log.Logf("Machine %s is up but still bootstrapping", thirdMachineReplacementName)
+
+		// The thirdMachine replacement is up, meaning that the test validated that remediation of the third CP machine works (note: this test remediation after the cluster is fully provisioned).
+
+		By("THIRD CP SUCCESSFULLY REMEDIATED!")
+
+		Byf("Unblock bootstrap for Machine %s and wait for it to be provisioned", thirdMachineReplacementName)
+		sendSignalToBootstrappingMachine(ctx, sendSignalToBootstrappingMachineInput{
+			client:    input.BootstrapClusterProxy.GetClient(),
+			namespace: namespace.Name,
+			machine:   thirdMachineReplacementName,
+			signal:    "pass",
+		})
+		log.Logf("Waiting for Machine %s to be provisioned", thirdMachineReplacementName)
+		Eventually(func() bool {
+			if err := input.BootstrapClusterProxy.GetClient().Get(ctx, client.ObjectKeyFromObject(thirdMachineReplacement), thirdMachineReplacement); err != nil {
+				return false
+			}
+			return thirdMachineReplacement.Status.NodeRef != nil
+		}, input.E2EConfig.GetIntervals(specName, "wait-machine-provisioned")...).Should(BeTrue(), "Machine %s failed to be provisioned", thirdMachineReplacementName)
+
+		// All three CP machines are up again.
+
+		By("CP BACK TO FULL OPERATIONAL STATE!")
+
+		By("PASSED!")
+	})
+
+	AfterEach(func() {
+		// Dumps all the resources in the spec namespace, then cleanups the cluster object and the spec namespace itself.
+		dumpSpecResourcesAndCleanup(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder, namespace, cancelWatches, clusterResources.Cluster, input.E2EConfig.GetIntervals, input.SkipCleanup)
+	})
+}
+
+func createConfigMapForMachinesBootstrapSignal(ctx context.Context, writer client.Writer, namespace string) {
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configMapName,
+			Namespace: namespace,
+		},
+		Data: map[string]string{
+			configMapDataKey: "hold",
+		},
+	}
+	Expect(writer.Create(ctx, cm)).To(Succeed(), "failed to create mhc-test config map")
+}
+
+type createWorkloadClusterAndWaitInput struct {
+	E2EConfig            *clusterctl.E2EConfig
+	clusterctlConfigPath string
+	proxy                framework.ClusterProxy
+	artifactFolder       string
+	specName             string
+	flavor               *string
+	namespace            string
+	authenticationToken  []byte
+	serverAddr           string
+}
+
+// createWorkloadClusterAndWait creates a workload cluster ard return ass soon as the cluster infrastructure is ready.
```

Comment:
```suggestion
// createWorkloadClusterAndWait creates a workload cluster ard returns as soon as the cluster infrastructure is ready.
```

Reply:
Should we add for future us that we have to do this because the func we normally use would fail if the control plane doesn't come up? (I assume that's the reason)

Reply:
Wondering if we should re-use the normal code. ~ split up a `ApplyClusterTemplate` func and use it here and in `ApplyClusterTemplateAndWait`)

Reply:
Added a comment, we can eventually iterate later if we want to make those use cases available to everyone through the framework, but for this first iteration this is a non-goal because I don't expect anyone outside the CAPI repository to run KCP remediation E2E tests.